src/Core/ConstraintValidator/ReductionValidator.php

```diff
@@ -0,0 +1,94 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\ConstraintValidator;
+
+use PrestaShop\PrestaShop\Core\ConstraintValidator\Constraints\Reduction as ReductionConstraint;
+use PrestaShop\PrestaShop\Core\Domain\Exception\DomainConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\ValueObject\Reduction;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+
+/**
+ * Validates reduction type and value
+ */
+final class ReductionValidator extends ConstraintValidator
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($value, Constraint $constraint)
+    {
+        if (!$constraint instanceof ReductionConstraint) {
+            throw new UnexpectedTypeException($constraint, ReductionConstraint::class);
+        }
+
+        if (null === $value || '' === $value) {
+            return;
+        }
+
+        if (!is_array($value)) {
+            throw new UnexpectedTypeException($value, 'array');
+        }
+
+        try {
+            new Reduction($value['type'], $value['value']);
```

Comment:
So you're using validation checks in `Reduction` that throw `DomainConstraintExceptions` and catch them in order to write the right error message and avoid breaking the flow ..?

Euh ... 🤔it works but it hijacks the standard workflow of these validation checks.

I've never seen this way of validating an object 😅 it might feel weird, but maybe it's right.
Need more opinion, ping @PierreRambaud @jolelievre @eternoendless @mickaelandrieu

Reply:
Look weird to use a constructor to validate data :thinking: 
ping @eternoendless

Reply:
I agree with @PierreRambaud, please move the validation logic outside of Reduction constructor 😉

Reply:
@matks although I agree that it's not the most convenient way to validate data, but it works.

We could of course duplicate validation that is inside `Reduction` constructor, but we can't move it out. Do you have anything else on your mind? :)